### PR TITLE
Story3

### PR DESCRIPTION
**Pull request by Henry-Tottle**

### CODE REVIEW

File: products.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
+<?php
+require_once ('vendor/autoload.php');
+use FurnitureStoreApi\Services\HeaderService;
+HeaderService::setHeaders();
```
Comment (matt-ioa) [inline products.php]: Put all the `use` statements together, then put this after

File: products.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+use FurnitureStoreApi\Products\ProductHydrator;
+use FurnitureStoreApi\Services\ResponseService;
+use FurnitureStoreApi\Services\DbConnection;
+
+$db = DbConnection::setConnection();
+$query = $db->prepare('SELECT MAX(`category_id`) FROM `products`');
+$query->execute();
+$maxCategory = $query->fetch();
+$maxCategory = $maxCategory[0];
```
Comment (matt-ioa) [inline products.php]: Seems like this could be part of your category hydrator

File: products.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+try
+{
+    $jsonString = [];
+    if ($_GET['cat']>$maxCategory)
+    {
+        ResponseService::makeResponse('Invalid category id', $jsonString, 400);
```
Comment (matt-ioa) [inline products.php]: Another way of doing this is to create your own custom exception type and throw it here: I can show you how to do this.

File: products.php [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        $jsonString = ProductHydrator::getProductsByCategory($_GET['cat']);
+        ResponseService::makeResponse('Successfully retrieved products', $jsonString, 200);
+    }
+}
+catch(Exception $exception)
+
+    {
+        ResponseService::makeResponse('Unexpected error', $jsonString, 400);
+    }
```
Comment (matt-ioa) [inline products.php:31]: If you create your own exception type, you can add another catch for that type here to handle it separately

**Review by matt-ioa — CHANGES_REQUESTED** [review]
A couple of changes needed, let me know if you want to implement and throw custom exceptions to handle 400 responses

File: categories.php [hunk 5]
```diff
@@ -4,6 +4,7 @@
 HeaderService::setHeaders();
 use \FurnitureStoreApi\Categories\CategoryHydrator;
 use FurnitureStoreApi\Services\ResponseService;
+use FurnitureStoreApi\Exceptions\CustomException;
```
Comment (matt-ioa) [inline categories.php]: Is this still needed?

File: products.php [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+use FurnitureStoreApi\Services\ResponseService;
+use FurnitureStoreApi\Services\DbConnection;
+use FurnitureStoreApi\Categories\CategoryHydrator;
+use FurnitureStoreApi\Exceptions\InvalidCategoryException;
+HeaderService::setHeaders();
+
+try
+{
+    $jsonString = [];
+    if ($_GET['cat']>CategoryHydrator::getMaxCategory())
```
Comment (matt-ioa) [inline products.php]: You could also check to see if the value < 0, or if it's not a number at all

File: products.php [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        $jsonString = ProductHydrator::getProductsByCategory($_GET['cat']);
+        ResponseService::makeResponse('Successfully retrieved products', $jsonString, 200);
+    }
+}
+catch (InvalidCategoryException $exception)
+{
+    ResponseService::makeResponse($exception->getMessage(), $jsonString, 400);
+}
+catch(Exception $exception)
+
```
Comment (matt-ioa) [inline products.php]: Remove this blank line

File: src/Categories/CategoryHydrator.php [hunk 8]
```diff
@@ -8,14 +8,24 @@ class CategoryHydrator
...
         return $query->fetchAll();
     }
+
+    public static function getMaxCategory()
+    {
+        $db = DbConnection::setConnection();
+        $query = $db->prepare('SELECT MAX(`category_id`) FROM `products`');
+        $query->execute();
+        $maxCategory = $query->fetch();
+        $maxCategory = $maxCategory[0];
```
Comment (matt-ioa) [inline src/Categories/CategoryHydrator.php]: Do you know why you have to do `[0]` here?

**Review by matt-ioa — CHANGES_REQUESTED** [review]
Just a couple of changes to make, take a look 👀

File: src/Categories/CategoryHydrator.php (hunk 8 above)
Comment (joseph-padfield) [reply src/Categories/CategoryHydrator.php]: It's the first item in the array

Comment (joseph-padfield) [reply src/Categories/CategoryHydrator.php]: (of 1 item!)

**Review by matt-ioa — APPROVED** [review]

**Pull request merged**